### PR TITLE
Add IISNode IICluster and Autoscaler classes

### PR DESCRIPTION
**Pull request by chrisgilbert**
I have been using the following to test.  

```
from tyr.clusters.iis import IISCluster
IISCluster(group='chrisgtest', environment='stage', ami='ami-bea09bd6', subnet_ids=["subnet-961652e1", "subnet-9a41d9c3"], desired_capacity=2, min_size=1, max_size=3).autorun()
```

There are a few issues that need ironing out:

- Problem spinning up servers once in ASG.  Need to diagnose

### CODE REVIEW

**Comment by jamiegs** [comment]
I don't think you'll need to worry about Classiclinking. We've got all of our prod web server in VPC, and we're probably not going to need to classic link when migrating or stage servers

File: tyr/policies/ec2.py [hunk 1]
```diff
@@ -84,3 +84,237 @@
...
+      "Effect": "Allow"
+    },
+    {
+      "Action": [
+        "cloudwatch:DescribeAlarms",
+        "cloudwatch:DescribeAlarmsForMetric",
+        "cloudwatch:ListMetrics",
+        "cloudwatch:PutMetricAlarm",
+        "cloudwatch:PutMetricData",
+        "cloudwatch:SetAlarmState"
```
Comment (wliao008) [inline tyr/policies/ec2.py]: would need access to `route53:ChangeResourceRecordSets` as well

File: tyr/policies/ec2.py [hunk 2]
```diff
@@ -84,3 +84,237 @@
...
```
Comment (wliao008) [inline tyr/policies/ec2.py]: missing `route53:ChangeResourceRecordSets`

File: tyr/servers/iis/node.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                                      availability_zone=availability_zone,
+                                      security_groups=security_groups,
+                                      subnet_id=subnet_id,
+                                      dns_zones=None,
+                                      add_route53_dns=False)
+
+        env_prefix = self.environment[0]
+
+        self.security_groups = [
+            "{0}-management".format(env_prefix),
```
Comment (wliao008) [inline tyr/servers/iis/node.py]: In vpc, the management security group is just `management` without the prefix.

File: tyr/autoscaler/autoscaler.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+import boto.ec2
+import logging
+
+
+class AutoScaler(object):
+    '''
+    Autoscaler setup class
+    '''
+
+    log = logging.getLogger('Clusters.AutoScaler')
```
Comment (citruspi) [inline tyr/autoscaler/autoscaler.py]: This module should either be moved to the `tyr.clusters` package or the logger name should be changed to correctly reflect the package it belongs to - perhaps `Autoscaling.Autoscaler` (if the package were renamed `autoscaling`).

File: tyr/autoscaler/autoscaler.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    log.setLevel(logging.DEBUG)
+    ch = logging.StreamHandler()
+    ch.setLevel(logging.DEBUG)
+    formatter = logging.Formatter(
+        '%(asctime)s [%(name)s] %(levelname)s: %(message)s',
+        datefmt='%H:%M:%S')
+    ch.setFormatter(formatter)
+    log.addHandler(ch)
```
Comment (citruspi) [inline tyr/autoscaler/autoscaler.py]: You're going to run into issues with multiple log handlers on the same logger. I'd recommend checking for existing handlers before adding a new one or, better yet, make use of the `establish_logger` method in `tyr.servers.server`.

File: tyr/autoscaler/autoscaler.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        self.max_size = max_size
+        self.min_size = min_size
+        self.default_cooldown = default_cooldown
+        self.health_check_grace_period = health_check_grace_period
+
+    def establish_autoscale_connection(self):
+        try:
+            self.conn = boto.ec2.autoscale.connect_to_region(self.node_obj.region)
+            self.log.info('Established connection to autoscale')
+        except:
```
Comment (citruspi) [inline tyr/autoscaler/autoscaler.py]: We normally try to [log exceptions][0] before raising them. (Note - that should probably actually be a `critical` level event).

[0]: https://github.com/hudl/Tyr/blob/master/tyr/servers/server.py#L617

File: tyr/autoscaler/autoscaler.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            raise
+
+    def create_launch_configuration(self):
+        self.log.info("Getting launch_configuration: {0}"
```
Comment (citruspi) [inline tyr/autoscaler/autoscaler.py]: Named placeholders are preferred to positional indexes.

File: tyr/autoscaler/autoscaler.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
+from boto.ec2.autoscale import AutoScaleConnection
```
Comment (citruspi) [inline tyr/autoscaler/autoscaler.py]: This is imported but never used.

File: tyr/autoscaler/autoscaler.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        if availability_zones:
+            self.autoscale_availability_zones = availability_zones
+        else:
+            self.autoscale_availability_zones = None
+
+        # If you set these they must match the availability zones
+        if subnet_ids:
+            self.autoscale_subnets = subnet_ids
+        else:
+            self.autoscale_subnets = None #(get the subnet IDs)
```
Comment (citruspi) [inline tyr/autoscaler/autoscaler.py]: Inline comments should have to spaces before the `#` and one after.

File: tyr/autoscaler/autoscaler.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+        if not subnet_ids and not availability_zones:
+            raise ValueError("Must specify either availability_zones or subnets.")
```
Comment (citruspi) [inline tyr/autoscaler/autoscaler.py]: This line is over 79 characters.

File: tyr/autoscaler/autoscaler.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+        self.node_obj = node_obj
+        self.max_size = max_size
+        self.min_size = min_size
+        self.default_cooldown = default_cooldown
+        self.health_check_grace_period = health_check_grace_period
+
+    def establish_autoscale_connection(self):
+        try:
+            self.conn = boto.ec2.autoscale.connect_to_region(self.node_obj.region)
```
Comment (citruspi) [inline tyr/autoscaler/autoscaler.py]: This line is over 79 characters.

File: tyr/autoscaler/autoscaler.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+            self.log.info('Established connection to autoscale')
+        except:
+            raise
+
+    def create_launch_configuration(self):
+        self.log.info("Getting launch_configuration: {0}"
+            .format(self.launch_configuration))
```
Comment (citruspi) [inline tyr/autoscaler/autoscaler.py]: `.format` should vertically align with the opening quote above it.

File: tyr/autoscaler/autoscaler.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+
+        lc = self.conn.get_all_launch_configurations(
+            names=[self.launch_configuration])
+        if not lc:
+            self.log.info("Creating new launch_configuration: {0}"
+                          .format(self.launch_configuration))
+            lc = LaunchConfiguration(name=self.launch_configuration,
+                                     image_id=self.node_obj.ami,
+                                     key_name=self.node_obj.keypair,
+                                     security_groups=self.node_obj.get_security_group_ids(
```
Comment (citruspi) [inline tyr/autoscaler/autoscaler.py]: This line is over 79 characters.

File: tyr/autoscaler/autoscaler.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    def create_autoscaling_group(self):
+        existing_asg = self.conn.get_all_groups(
+            names=[self.autoscaling_group])
+
+        if not existing_asg:
+            self.log.info("Creating new autoscaling group: {0}"
+                          .format(self.autoscaling_group))
+
+            ag = AutoScalingGroup(name=self.autoscaling_group,
+                                  availability_zones=self.autoscale_availability_zones,
```
Comment (citruspi) [inline tyr/autoscaler/autoscaler.py]: This line is over 79 characters.

File: tyr/autoscaler/autoscaler.py [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+                                  desired_capacity=self.desired_capacity,
+                                  health_check_period=self.health_check_grace_period,
```
Comment (citruspi) [inline tyr/autoscaler/autoscaler.py]: This line is over 79 characters.

File: tyr/clusters/iis.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+
+        auto = AutoScaler(launch_configuration=self.launch_configuration,
+                          autoscaling_group=self.autoscaling_group,
+                          desired_capacity=self.desired_capacity,
+                          max_size=self.max_size,
+                          min_size=self.min_size,
+                          default_cooldown=self.default_cooldown,
+                          availability_zones=self.availability_zones,
+                          subnet_ids=self.subnet_ids,
+                          health_check_grace_period=self.health_check_grace_period,
```
Comment (citruspi) [inline tyr/clusters/iis.py]: This line is over 79 characters.

File: tyr/servers/iis/node.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+            mongo_ops = ("MongosHost", "Disabled", "MongosService")
+            if self.mongos_service not in mongo_ops:
+                raise ValueError(
+                    "Mongo service name must be one of: {0}".format(
+                        mongo_ops))
+
+        self.ports_to_authorize = [9000, 9001, 8095, 8096]
+
+        self.IAM_ROLE_POLICIES.append('allow-web-initialization-{0}'
+            .format(self.environment))
```
Comment (citruspi) [inline tyr/servers/iis/node.py]: `.format` should be indented further to align with the opening quote of the string above it.

File: tyr/servers/iis/node.py [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+        self.IAM_ROLE_POLICIES.append('allow-outpost-sns-{0}'
+            .format(self.environment))
```
Comment (citruspi) [inline tyr/servers/iis/node.py]: `.format` should be indented further to align with the opening quote of the string above it.

File: tyr/servers/server.py [hunk 19]
```diff
@@ -790,6 +800,25 @@ def route(self, wait=False):
                     self.log.error(str(e))
                     raise e
 
+    def ingress_rules(self):
```
Comment (citruspi) [inline tyr/servers/server.py]: Is this just adding ingress security group rules? Is it something that could be done with #32?

Comment (chrisgilbert) [inline tyr/servers/server.py]: Yes, that's right, it could use the new code.  Is that ready for a merge?

**Comment by chrisgilbert** [comment]
OK, I have made changes as discussed.

**Comment by chrisgilbert** [comment]
Meh, my revert didn't reopen the pull request.  Not sure what's happened there.  I will create a new pull request for these changes

**Pull request merged**